src/main/java/at/ac/tuwien/kr/alpha/grounder/transformation/ArithmeticTermsRewriting.java

```diff
@@ -0,0 +1,179 @@
+			}
+		}
+		// Rewrite body.
+		for (Literal literal : inputProgramRule.getBody()) {
+			if (!containsArithmeticTermsToRewrite(literal.getAtom())) {
+				// Keep body literal as-is if no ArithmeticTerm occurs.
+				rewrittenBodyLiterals.add(literal);
+				continue;
+			}
+			rewrittenBodyLiterals.add(rewriteAtom(literal.getAtom(), rewrittenBodyLiterals).toLiteral(!literal.isNegated()));
+		}
+		return new NormalRule(rewrittenHead, rewrittenBodyLiterals);
+	}
+
+	private boolean containsArithmeticTermsToRewrite(NormalRule inputProgramRule) {
+		if (!inputProgramRule.isConstraint()) {
+			Atom headAtom = inputProgramRule.getHeadAtom();
+			if (containsArithmeticTermsToRewrite(headAtom)) {
+				return true;
+			}
+		}
+		// Check whether body contains an ArithmeticTerm.
+		for (Literal literal : inputProgramRule.getBody()) {
+			if (containsArithmeticTermsToRewrite(literal.getAtom())) {
+				return true;
+			}
+		}
+		return false;
+	}
+
+	private Term rewriteArithmeticSubterms(Term term, List<Literal> bodyLiterals) {
+		// Keep term as-is if it contains no ArithmeticTerm.
+		if (!containsArithmeticTerm(term)) {
+			return term;
+		}
+		// Switch on term type.
+		if (term instanceof ArithmeticTerm) {
+			VariableTerm replacementVariable = VariableTerm.getInstance(ARITHMETIC_VARIABLES_PREFIX + numArithmeticVariables++);
+			bodyLiterals.add(new ComparisonAtom(replacementVariable, term, ComparisonOperator.EQ).toLiteral());
+			return replacementVariable;
+		} else if (term instanceof VariableTerm || term instanceof ConstantTerm) {
+			return term;
+		} else if (term instanceof FunctionTerm) {
+			List<Term> termList = ((FunctionTerm) term).getTerms();
+			List<Term> rewrittenTermList = new ArrayList<>();
+			for (Term subterm : termList) {
+				rewrittenTermList.add(rewriteArithmeticSubterms(subterm, bodyLiterals));
+			}
+			return FunctionTerm.getInstance(((FunctionTerm) term).getSymbol(), rewrittenTermList);
+		} else {
+			throw oops("Rewriting unknown Term type: " + term.getClass());
+		}
+	}
+
+	private Atom rewriteAtom(Atom atomToRewrite, List<Literal> bodyLiterals) {
+		List<Term> rewrittenTerms = new ArrayList<>();
+		for (Term atomTerm : atomToRewrite.getTerms()) {
+			// Rewrite arithmetic term.
+			rewrittenTerms.add(rewriteArithmeticSubterms(atomTerm, bodyLiterals));
+		}
+
+		// NOTE: we have to construct an atom of the same type as atomToRewrite, there should be a nicer way than that instanceof checks below.
+		if (atomToRewrite instanceof BasicAtom) {
+			return new BasicAtom(atomToRewrite.getPredicate(), rewrittenTerms);
+		} else if (atomToRewrite instanceof ComparisonAtom) {
+			throw oops("Trying to rewrite ComparisonAtom.");
+		} else if (atomToRewrite instanceof ExternalAtom) {
+			// Rewrite output terms, as so-far only the input terms list has been rewritten.
+			List<Term> rewrittenOutputTerms = new ArrayList<>();
+			for (Term term : ((ExternalAtom) atomToRewrite).getOutput()) {
```

Comment:
Codecov is complaining that the `ExternalAtom` part is not covered by any tests. Generally, I think it would make sense to have a dedicated `ArithmeticTermsRewritingTest` in addition to the existing (more "regession-test-like") `ArithmeticTermsTest`. The advantage of such a "real unit test" would be (in addition to codecov being able to measure coverage better) that a set of more fine-grained tests would also be an ideal starting point for debugging if ever any issues related to this code occurred.

Reply:
`ArithmeticTermsRewritingTest` is now also included.